src/library/scala/collection/mutable/RedBlackTree.scala

```diff
@@ -0,0 +1,502 @@
+  }
+
+  object Tree {
+    def empty[A, B]: Tree[A, B] = new Tree(null, 0)
+  }
+
+  object Node {
+
+    @inline def apply[A, B](key: A, value: B, red: Boolean,
+                            left: Node[A, B], right: Node[A, B], parent: Node[A, B]): Node[A, B] =
+      new Node(key, value, red, left, right, parent)
+
+    @inline def leaf[A, B](key: A, value: B, red: Boolean, parent: Node[A, B]): Node[A, B] =
+      new Node(key, value, red, null, null, parent)
+
+    def unapply[A, B](t: Node[A, B]) = Some((t.key, t.value, t.left, t.right, t.parent))
+  }
+
+  // ---- getters ----
+
+  def isRed(node: Node[_, _]) = (node ne null) && node.red
+  def isBlack(node: Node[_, _]) = (node eq null) || !node.red
+
+  // ---- size ----
+
+  def size(node: Node[_, _]): Int = if (node eq null) 0 else 1 + size(node.left) + size(node.right)
+  def isEmpty(tree: Tree[_, _]) = tree.root eq null
+
+  // ---- search ----
+
+  def get[A: Ordering, B](tree: Tree[A, B], key: A): Option[B] = getNode(tree.root, key) match {
+    case null => None
+    case node => Some(node.value)
+  }
+
+  @tailrec private[this] def getNode[A, B](node: Node[A, B], key: A)(implicit ord: Ordering[A]): Node[A, B] =
+    if (node eq null) null
+    else {
+      val cmp = ord.compare(key, node.key)
+      if (cmp < 0) getNode(node.left, key)
+      else if (cmp > 0) getNode(node.right, key)
+      else node
+    }
+
+  def contains[A: Ordering](tree: Tree[A, _], key: A) = getNode(tree.root, key) ne null
+
+  def min[A, B](tree: Tree[A, B]): Option[(A, B)] = minNode(tree.root) match {
+    case null => None
+    case node => Some((node.key, node.value))
+  }
+
+  private def minNode[A, B](node: Node[A, B]): Node[A, B] =
+    if (node eq null) null else minNodeNonNull(node)
+
+  @tailrec def minNodeNonNull[A, B](node: Node[A, B]): Node[A, B] =
+    if (node.left eq null) node else minNodeNonNull(node.left)
+
+  def max[A, B](tree: Tree[A, B]): Option[(A, B)] = maxNode(tree.root) match {
+    case null => None
+    case node => Some((node.key, node.value))
+  }
+
+  private def maxNode[A, B](node: Node[A, B]): Node[A, B] =
+    if (node eq null) null else maxNodeNonNull(node)
+
+  @tailrec def maxNodeNonNull[A, B](node: Node[A, B]): Node[A, B] =
+    if (node.right eq null) node else maxNodeNonNull(node.right)
+
+  def minAfter[A, B](tree: Tree[A, B], key: A)(implicit ord: Ordering[A]): Option[(A, B)] =
+    minNodeAfter(tree.root, key) match {
+      case null => None
+      case node => Some((node.key, node.value))
+    }
+
+  private[this] def minNodeAfter[A, B](node: Node[A, B], key: A)(implicit ord: Ordering[A]): Node[A, B] = {
+    if (node eq null) null
+    else {
+      var y: Node[A, B] = null
+      var x = node
+      var cmp = 1
+      while ((x ne null) && cmp != 0) {
+        y = x
+        cmp = ord.compare(key, x.key)
+        x = if (cmp < 0) x.left else x.right
+      }
+      if (cmp <= 0) y else successor(y)
```

Comment:
Yes, `minAfter` and `maxBefore` follow the semantics of `from` and `until` in ranges. I agree that it may be misleading, I'll add a comment explaining that.